Model coloring

lilleyse opened the pull request:
I was inspired by [this post on the forum](https://groups.google.com/forum/#!topic/cesium-dev/vtUKkCasLBM) to be able to color a model with a solid color.

The blend color is mixed with the rendered colored. It's not possible to just replace the diffuse color unless some semantic is defined, but this is probably not worth the trouble.

Is this something we want to support in the Entity API as well?

![truck](https://cloud.githubusercontent.com/assets/915398/19826349/b4ff0e4e-9d55-11e6-8c01-f3a5680ee2de.png)

![airplane](https://cloud.githubusercontent.com/assets/915398/19826306/d8420ace-9d54-11e6-8142-f8f274a06127.png)

pjcozzi commented:
Somewhat related to #2387 and #4314.

pjcozzi commented on Source/Scene/Model.js:
```diff
@@ -517,6 +521,24 @@ define([
+         *
+         * @default Color.RED
+         */
+        this.blendColor = defaultValue(options.blendColor, Color.RED);
```
Is the intention to allow RGBA values?  If so, the render state may need to change on the fly when alpha changes from one to non-one.  Likewise, you may want to optimize the case when alpha is zero and not issue the `DrawCommand`.

If the intention is only RGB, the explicitly document that and throw when alpha is non-zero.

I would support alpha so users can fade in/out a model with this.

lilleyse replied:
I had intentionally left alpha out, but I do think it will be nice to support in this PR.

pjcozzi commented on Source/Scene/Model.js:
```diff
@@ -1719,6 +1741,19 @@ define([
+    function modifyShaderForBlendColor(shader) {
+        shader = ShaderSource.replaceMain(shader, 'czm_blend_main');
+        shader +=
+            'uniform vec4 czm_blendColor; \n' +
```
Instead of `czm_`, should we use something like `gltf_` (like we use `tile_` in 3D Tiles)?  I probably used `czm_` elsewhere for 3D models, but perhaps we should reserve it for built-ins.

pjcozzi commented:
Add this to the Sandcastle example.

And yes, I expect we would want Entity API support and eventually CZML.

lilleyse commented:
Ready for review.

Added Entity API and CZML support. Also added alpha support.

![alpha](https://cloud.githubusercontent.com/assets/915398/20904029/c4843bae-bb0b-11e6-8346-41db63621c51.JPG)

lilleyse commented:
There are still some unavoidable artifacts when OIT is off due to lack of triangle sorting. The back/front commands help though, but can't prevent multiple front faces from being drawn in arbitrary order.

This is with alpha of 0.9 to highlight the issue.
![artifacts](https://cloud.githubusercontent.com/assets/915398/20904123/2889001c-bb0c-11e6-9d7f-1a87b6131bc5.JPG)

lilleyse commented:
Another note: glTF that are marked as premultiplied are now un-premultiplied. Besides making it easier to blend colors in this PR, it will also improve OIT which overrides the glTF's render state and always does un-premultiplied alpha blending.

pjcozzi commented:
> The back/front commands help though, but can't prevent multiple front faces from being drawn in arbitrary order.

Ah yes, this will only work for convex objects.  Perhaps it is not worth it since this is a fallback and only a partial solution that is unlikely to be enough in common cases.

lilleyse commented:
> Ah yes, this will only work for convex objects. Perhaps it is not worth it since this is a fallback and only a partial solution that is unlikely to be enough in common cases.

Ok, then it may be worth changing 3d-tiles as well which also generates translucent front/back commands.

pjcozzi commented:
Is Blend Alpha supposed to work when Blend Amount is zero?  Blend Color does not.

![ok](https://cloud.githubusercontent.com/assets/782098/20904594/4159985c-bb0e-11e6-9b03-eac44f9dddf3.gif)

pjcozzi commented:
In the Sandcastle example, Blend Color going from 0 to 1 is awkward.  If a color picker is trivial, let's use it; otherwise, anything else like a dropdown would be fine.

![ok](https://cloud.githubusercontent.com/assets/782098/20904650/7d638934-bb0e-11e6-837d-3921a6b501e6.gif)

pjcozzi commented:
> Ok, then it may be worth changing 3d-tiles as well which also generates translucent front/back commands.

OK with me, probably my doing.

lilleyse commented:
> Is Blend Alpha supposed to work when Blend Amount is zero? Blend Color does not.

Yes this could be documented better, but it is not tied to Blend Amount because that will change the color in the process. Someone may want to only change the alpha portion.

pjcozzi commented on Apps/Sandcastle/gallery/3D Models.html:
```diff
@@ -43,14 +111,16 @@
-            maximumScale : 20000
+            maximumScale : 20000,
+            blendAmount : viewModel.blendAmount,
+            blendColor : Cesium.Color.fromHsl(viewModel.blendColor, 1.0, 0.5, viewModel.blendAlpha)
```
Do our users think of this as blending?  Or just as applying a color to a model like [Billboard.color](https://github.com/AnalyticalGraphicsInc/cesium/blob/master/Source/Scene/Billboard.js#L592)?

What do you think of renaming this to `color`?  I'm not sure what to rename `blendAmount` to...is there anywhere else in the Cesium API to mirror?

pjcozzi commented:
> Yes this could be documented better, but it is not tied to Blend Amount because that will change the color in the process. Someone may want to only change the alpha portion.

Couldn't they just set the color to white?

lilleyse commented:
> Couldn't they just set the color to white?

Not currently because the the model coloring is doing a mix, not a multiply.

pjcozzi commented:
>>Couldn't they just set the color to white?
>Not currently because the the model coloring is doing a mix, not a multiply.

I don't want to make this too complicated, but should we follow the same pattern as 3D Tiles features?  Do you have use cases for mix?  There will be lots of use cases for multiply, e.g., blue/red force.

lilleyse commented:
> I don't want to make this too complicated, but should we follow the same pattern as 3D Tiles features? Do you have use cases for mix? There will be lots of use cases for multiply, e.g., blue/red force.

Yeah I think supporting both modes may be a good idea, similar to `Cesium3DTileColorBlendMode`. Mix is nice for models that aren't close to white. Like a green model with a red `blendColor` will be black with multiply.

pjcozzi commented:
Looks good, just those comments.

lilleyse commented on Source/Scene/Model.js:
```diff
@@ -3386,6 +3470,43 @@ define([
+        return translucentCommand;
+    }
+
+    function updateBlendColor(model, frameState) {
```
Yes both this and the 2D commands can be derived commands. I'll wait for #4314.

lilleyse commented:
Ready now. I borrowed `ColorBlendMode` from 3d-tiles so HIGHLIGHT, REPLACE, and MIX are supported.

pjcozzi commented:
Is the separate Firefox issue, #4658, right?

![ok](https://cloud.githubusercontent.com/assets/782098/20930516/9ef5822c-bb9b-11e6-89db-2ce352b28135.gif)

pjcozzi commented on Source/DataSources/CzmlDataSource.js:
```diff
@@ -34,6 +34,7 @@ define([
         '../Core/Spherical',
         '../Core/TimeInterval',
         '../Core/TimeIntervalCollection',
+        '../Scene/ColorBlendMode',
```
Can you update the CZML doc or submit an issue to update it?

lilleyse replied:
Will do

pjcozzi commented on Source/Scene/Model.js:
```diff
@@ -2839,6 +2919,27 @@ define([
+        };
+    }
+
+    function createColorBlendFunction(model) {
```
Does 3D Tiles also use this?  If so, should we make this a utility function on `ColorBlendMode` that takes a `blendAmount`?

pjcozzi commented:
That's all my comments.  This is a masterpiece.

lilleyse commented on Apps/Sandcastle/gallery/3D Models.html:
```diff
@@ -19,10 +19,45 @@
 <body class="sandcastle-loading" data-sandcastle-bucket="bucket-requirejs.html">
 <style>
     @import url(../templates/bucket.css);
+    #toolbar {
```
Plus silhouettes can be combined in that new demo.

lilleyse commented:
Ready

lilleyse commented:
Just have one small case to add when the alpha is 0.

lilleyse commented:
Ready now.

>Is the separate Firefox issue, #4658, right?

Yeah

lilleyse commented:
CZML doc PR here: https://github.com/AnalyticalGraphicsInc/czml-writer/pull/122
